From be8e1c3103e57da9a9c2f72d79e69194c1f1d0b9 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Thu, 26 Apr 2012 06:12:30 +0000
Subject: [PATCH] [CONFIGURATION-495] HierarchicalConfiguration.setProperty()
 now handles lists and arrays correctly if delimiter parsing is disabled.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1330666 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                       |  4 ++
 .../HierarchicalConfiguration.java            |  2 +-
 .../TestPropertiesConfiguration.java          | 20 ++++++++++
 .../configuration/TestXMLConfiguration.java   | 39 +++++++++++++++++++
 4 files changed, 64 insertions(+), 1 deletion(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index aef9eef6e1..1172aa875e 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -27,6 +27,10 @@
   <body>
     <release version="1.9" date="in SVN"
       description="TBD">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-495">
+        List properties can now be set correctly on a HierarchicalConfiguration
+        if delimiter parsing is disabled.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-487">
         DataConfiguration.get() now also works with String properties and if no
         data type conversion is required.
diff --git a/src/main/java/org/apache/commons/configuration/HierarchicalConfiguration.java b/src/main/java/org/apache/commons/configuration/HierarchicalConfiguration.java
index 149161de45..527deeb9a2 100644
--- a/src/main/java/org/apache/commons/configuration/HierarchicalConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration/HierarchicalConfiguration.java
@@ -739,7 +739,7 @@ public void setProperty(String key, Object value)
         // Update the existing nodes for this property
         Iterator<ConfigurationNode> itNodes = fetchNodeList(key).iterator();
         Iterator<?> itValues;
-        if (!isDelimiterParsingDisabled())
+        if (!isDelimiterParsingDisabled() || !(value instanceof String))
         {
             itValues = PropertyConverter.toIterator(value, getListDelimiter());
         }
diff --git a/src/test/java/org/apache/commons/configuration/TestPropertiesConfiguration.java b/src/test/java/org/apache/commons/configuration/TestPropertiesConfiguration.java
index bc15d96577..0cf6873220 100644
--- a/src/test/java/org/apache/commons/configuration/TestPropertiesConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration/TestPropertiesConfiguration.java
@@ -44,6 +44,7 @@
 import java.net.URLConnection;
 import java.net.URLStreamHandler;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
@@ -1057,6 +1058,25 @@ public void testBackslashEscapingInLists() throws Exception
         checkBackslashList("share1");
     }
 
+    /**
+     * Tests whether a list property is handled correctly if delimiter parsing
+     * is disabled. This test is related to CONFIGURATION-495.
+     */
+    @Test
+    public void testSetPropertyListWithDelimiterParsingDisabled()
+            throws ConfigurationException
+    {
+        String prop = "delimiterListProp";
+        conf.setDelimiterParsingDisabled(true);
+        List<String> list = Arrays.asList("val", "val2", "val3");
+        conf.setProperty(prop, list);
+        conf.setFile(testSavePropertiesFile);
+        conf.save();
+        conf.clear();
+        conf.load();
+        assertEquals("Wrong list property", list, conf.getProperty(prop));
+    }
+
     /**
      * Helper method for testing the content of a list with elements that
      * contain backslashes.
diff --git a/src/test/java/org/apache/commons/configuration/TestXMLConfiguration.java b/src/test/java/org/apache/commons/configuration/TestXMLConfiguration.java
index a8435b057d..afcf76269a 100644
--- a/src/test/java/org/apache/commons/configuration/TestXMLConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration/TestXMLConfiguration.java
@@ -34,6 +34,7 @@
 import java.io.StringWriter;
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
@@ -1848,6 +1849,44 @@ public void testAddNodesToSubnodeConfiguration() throws Exception
                 newNode instanceof XMLConfiguration.XMLNode);
     }
 
+    /**
+     * Tests whether list properties are set correctly if delimiter
+     * parsing is disabled. This test is related to CONFIGURATION-495.
+     */
+    @Test
+    public void testSetPropertyListWithDelimiterParsingDisabled()
+            throws ConfigurationException
+    {
+        String prop = "delimiterListProp";
+        conf.setDelimiterParsingDisabled(true);
+        List<String> list = Arrays.asList("val", "val2", "val3");
+        conf.setProperty(prop, list);
+        conf.setFile(testSaveFile);
+        conf.save();
+        conf.clear();
+        conf.load();
+        assertEquals("Wrong list property", list, conf.getProperty(prop));
+    }
+
+    /**
+     * Tests whether list properties are added correctly if delimiter parsing is
+     * disabled. This test is related to CONFIGURATION-495.
+     */
+    @Test
+    public void testAddPropertyListWithDelimiterParsingDisabled()
+            throws ConfigurationException
+    {
+        String prop = "delimiterListProp";
+        conf.setDelimiterParsingDisabled(true);
+        List<String> list = Arrays.asList("val", "val2", "val3");
+        conf.addProperty(prop, list);
+        conf.setFile(testSaveFile);
+        conf.save();
+        conf.clear();
+        conf.load();
+        assertEquals("Wrong list property", list, conf.getProperty(prop));
+    }
+
     /**
      * Prepares a configuration object for testing a reload operation.
      *
